Expand `storage.modifier.php` specificity across grammar

alexchexes opened the pull request:
## Description of the Change

This PR adds more specific scopes to existing `storage.modifier.php` tokens, so grammar consumers can target those modifiers more precisely.

It keeps backward compatibility by preserving `storage.modifier.php` on affected tokens, so those tokens now have scopes like `storage.modifier.php, storage.modifier.visibility.public.php`, etc. This is done via capture stacking (overlapping captures), to ensure broader compatibility with grammar consumers beyond VS Code.

It applies a consistent dual-scoping model to all relevant places where `storage.modifier.php` tokens are emitted, including PHPDoc `@access` values, the `global` modifier, and enum `case`.

### Covered contexts:

- Class modifiers: `final`, `abstract`, `readonly` (including anonymous class `readonly`).
- Constructor/method modifiers: `final`, `abstract`, `public`, `private`, `protected`, `static`.
- Property and promoted-property modifiers: `public`, `private`, `protected`, `readonly`, `static`.
- Asymmetric visibility tokens: `public(set)`, `private(set)`, `protected(set)` mapped to:
  - `storage.modifier.visibility.set.public.php`
  - `storage.modifier.visibility.set.private.php`
  - `storage.modifier.visibility.set.protected.php`
- Visibility bucket:
  - `storage.modifier.visibility.public.php`
  - `storage.modifier.visibility.private.php`
  - `storage.modifier.visibility.protected.php`
- Inheritance bucket:
  - `storage.modifier.inheritance.abstract.php`
  - `storage.modifier.inheritance.final.php`
- `readonly` and `static` are standalone:
  - `storage.modifier.readonly.php`
  - `storage.modifier.static.php`
- Trait adaptation alias modifiers in `use ... { ... as ...; }` (`as private`, `as static`, `as final`, etc.), while keeping non-modifier alias fallback as `entity.other.alias.php`. This also fixes an existing issue where `private/static/final` in `as private/static/final` (without an alias name) were incorrectly receiving `entity.other.alias.php` instead of `storage.modifier.php`.
- `global` keyword where it is currently treated as `storage.modifier.php`.
- Enum declaration `case` (enum body) with specific `storage.modifier.case.php`.
- PHPDoc `@access public|private|protected` value token with generic + specific scopes; invalid values remain invalid-scoped.

No other behavior changes outside scope specificity expansion:
- `static` return type remains `storage.type.php`.
- `switch case` remains `keyword.control.case.php`.

## Alternate Designs

### From the grammar syntax perspective:
Instead of the current capture-group-based approach, I considered using space-separated scope-name stacking in one `name` field.
Rejected to avoid relying on uncertain support from grammar consumers.

### From the consumer perspective:
Considered adding only buckets, i.e. `storage.modifier.visibility.php` for public/private/protected/(set).
Rejected to allow grammar consumers to target modifiers more specifically.

## Possible Drawbacks

Backward compatibility is preserved by leaving existing `storage.modifier.php` on those modifiers as is, so nothing breaks for existing grammar consumers.
The main drawback is mostly aesthetic: this BC plumbing makes token output more verbose (for example `storage.modifier.php storage.modifier.public.set.php`, where it could probably be just `storage.modifier.public.set.php`).

## Applicable Issues

None

## Screenshots of possible theme output:

Before/After:

<img width="350" alt="2026-03-10--18-04-22--033" src="https://github.com/user-attachments/assets/f2e1b54f-3d9b-45ee-9058-f3628333bfa9" />
<img width="352" alt="2026-03-10--18-03-42--250" src="https://github.com/user-attachments/assets/f39c8b76-827a-4d34-908a-8fd86540d09d" />


Clear distinction between modifiers of different kinds becomes possible.

alexchexes committed:
Add unique scope for "static" keyword

- Split the existing regular expression patterns to separate the "static" keyword.
- Added a unique scope "storage.modifier.static.php"

alexchexes commented:
@KapitanOczywisty May I ask if this PR hasn't moved forward because it lacked tests or for some other reason? I see it now has conflicts with the master branch, but it's been 3 years, so before resolving them I'd like to understand whether that's what's preventing this PR from progressing further.

KapitanOczywisty commented:
I'm not working for the Microsoft and they were not interested at all about maintaining this repository or even having someone with write access, so you're relying on my free time, which is very sparse in recent years.

The main thing is that many templates are relying on `storage.modifier.php`, so I'm a bit uneasy about changing that.

KapitanOczywisty commented on grammars/php.cson:
Unless it is necesary I'd keep single match, and tested for patter in capture. (Sorry for broken tabulation, GH still sucks...)
```suggestion
              \\s+(final|abstract|public|private|protected|static)
             \\s+([a-z_\\x{7f}-\\x{10ffff}][a-z0-9_\\x{7f}-\\x{10ffff}]*)
           '''
            'captures':
            '1':
            'name': 'keyword.other.use-as.php'
            '2':
'patterns': [
          {
            'match': 'static'
            'name': 'storage.modifier.static.php'
          }
          {
            'match': 'final|abstract|public|private|protected'
            'name': 'storage.modifier.php'
          }
        ]
            '3':
            'name': 'entity.other.alias.php'
            }
```

KapitanOczywisty committed:
Merge branch 'master' into patch-1

alexchexes commented:
Thanks, I wrote that follow-up yesterday and then realized that the problem we'd like to address is a bit broader than just "static" missing a distinct scope. If you don't mind, see this issue I created for myself and tell me if you're okay with that direction: https://github.com/alexchexes/language-php/issues/2

If it's ok, I'll start thinking about the best approach to address it, then later add a new commit to this PR.

KapitanOczywisty commented:
Not ideal, but a bit better could be double-scoping these e.g. `storage.modifier.php, storage.modifier.final.php`. With pattern for the most of them `storage.modifier.$1.php`, or for set ones: `(public|protected|private)\(set\)` => `storage.modifier.$1.set.php`

alexchexes committed:
Expand `storage.modifier.php` specificity across grammar

alexchexes committed:
Merge branch 'master' into patch-1

alexchexes commented:
@KapitanOczywisty I implemented this as you suggested and updated the PR description, please take a look.

I used capture stacking to add a second scope, without using space-separated stacking, since support for that across consumers is still unclear to me.

As an afterthought, I wondered whether it might be worth adding buckets (like `visibility` for public/private/protected, etc). That way, it could look like:

`storage.modifier.visibility.public.php`
`storage.modifier.visibility.private.php`
`storage.modifier.visibility.protected.php`
`storage.modifier.visibility.asymmetric.public.php` (for `(set)`)
`storage.modifier.visibility.asymmetric.private.php` (for `(set)`)
`storage.modifier.visibility.asymmetric.protected.php` (for `(set)`)
`storage.modifier.inheritance.abstract.php`
`storage.modifier.inheritance.final.php`

and others in a generic `declaration` bucket, like:

`storage.modifier.declaration.readonly.php`
`storage.modifier.declaration.static.php`

or, if we foresee more modifiers in the future, we could add them to provisional buckets like:

`storage.modifier.mutability.readonly.php`
`storage.modifier.storage.static.php`

What do you think about this?

KapitanOczywisty commented:
With bucket this is more convincing, since you can style whole buckets.

```
storage.modifier.visibility.public.php
storage.modifier.visibility.private.php
storage.modifier.visibility.protected.php
```
Asymmetric visibility means 2 different modifier not set in particular, thus `set` should be the common part for `(set)` ones.
```
storage.modifier.visibility.set.public.php or even storage.modifier.visibility.public.set.php
storage.modifier.visibility.set.private.php
storage.modifier.visibility.set.protected.php
```
Probably ok:
```
storage.modifier.inheritance.abstract.php
storage.modifier.inheritance.final.php
```
These are unlikely to have other counterparts, thus they could even be their own type.
```
storage.modifier.declaration.readonly.php or storage.modifier.readonly.php
storage.modifier.declaration.static.php or storage.modifier.static.php
```

alexchexes committed:
add buckets

alexchexes commented:
@KapitanOczywisty Done.
For asymmetric visibility modifiers, I took your first suggested format
`storage.modifier.visibility.set.*.php`
and not
`storage.modifier.visibility.*.set.php`

Because that allows targeting the asymmetric bucket with the selector `storage.modifier.visibility.set`.

And the buckets
`visibility.*` + `visibility.set.*`
feel like a more natural grouping than
`visibility.public.*` + `visibility.private.*` + `visibility.protected.*`.

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -264,11 +287,13 @@
               '1':
                 'name': 'storage.modifier.php'
               '2':
+                'name': 'storage.modifier.case.php'
```
I know I did write `storage.modifier`, however even then I wasn't sure what was proper class for case in enums, so if you have better idea this would be time to change this.

alexchexes replied:
The `storage.type.*` prefix seems like a better fit: `const` already uses `storage.type.const.php`, and enum `case` appears in a similar declaration context to `const` (and enums can also contain `const` declarations alongside `case` declarations).

Themes also often have broad `storage.type` selector styling, so this is less likely to leave enum `case` unstyled than introducing a completely new top-level scope family.

So we could use either `storage.type.case.php` or `storage.type.enum.case.php`. My vote is for the latter because:
1. It is more explicit and semantically clearer.
2. We already have `storage.type.enum.php` for the `enum` keyword itself, so themes can target both `enum` and enum `case` with the same `storage.type.enum` selector if they want to.

KapitanOczywisty replied:
```suggestion
                'name': 'storage.type.enum.case.php'
```

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -285,19 +310,20 @@
+              '2':
+                'name': 'storage.modifier.inheritance.${2:/downcase}.php'
+              '3':
+                'name': 'storage.modifier.readonly.php'
```
```suggestion
            'match': '(?i)(final|abstract)|(readonly)'
            'name': 'storage.modifier.php'
            'captures':
              '1':
                'name': 'storage.modifier.inheritance.${1:/downcase}.php'
              '2':
                'name': 'storage.modifier.readonly.php'
```

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -609,8 +637,13 @@
+              '2':
+                'name': 'storage.modifier.inheritance.${2:/downcase}.php'
+              '3':
+                'name': 'storage.modifier.visibility.${3:/downcase}.php'
```
```suggestion
            'match': '(?i)(final|abstract)|(public|private|protected)'
            'name': 'storage.modifier.php'
            'captures':
              '1':
                'name': 'storage.modifier.inheritance.${1:/downcase}.php'
              '2':
                'name': 'storage.modifier.visibility.${2:/downcase}.php'
```

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -659,8 +692,22 @@
+                  '4':
+                    'name': 'storage.modifier.visibility.${4:/downcase}.php'
+                  '5':
+                    'name': 'storage.modifier.readonly.php'
```
```suggestion
                'match': '''(?xi)
                  \\b(?:
                  ((public|private|protected)\\(set\\))
                  |(public|private|protected)
                  |(readonly)
                  )\\b
                '''
                'name': 'storage.modifier.php'
                'captures':
                  '1':
                    'name': 'storage.modifier.visibility.set.${2:/downcase}.php'
                  '3':
                    'name': 'storage.modifier.visibility.${3:/downcase}.php'
                  '4':
                    'name': 'storage.modifier.readonly.php'
```

alexchexes replied:
That cannot match `...(set)`, because there is no word boundary after `)`; so `\b` should not be added after `(set)`

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -713,8 +760,21 @@
+              '3':
+                'name': 'storage.modifier.visibility.${3:/downcase}.php'
+              '4':
+                'name': 'storage.modifier.static.php'
```
```suggestion
            'match': '''(?xi)
            (final|abstract)
            |(public|private|protected)
            |(static)
            '''
            'name': 'storage.modifier.php'
            'captures':
              '1':
                'name': 'storage.modifier.inheritance.${1:/downcase}.php'
              '2':
                'name': 'storage.modifier.visibility.${2:/downcase}.php'
              '3':
                'name': 'storage.modifier.static.php'
```

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -785,8 +845,22 @@
+              '4':
+                'name': 'storage.modifier.visibility.${4:/downcase}.php'
+              '5':
+                'name': 'storage.modifier.${5:/downcase}.php'
```
```suggestion
                'match': '''(?xi)
                  \\b(?:
                  ((public|private|protected)\\(set\\))
                  |(public|private|protected)
                  |(readonly|static)
                  )\\b
                '''
                'name': 'storage.modifier.php'
                'captures':
                  '1':
                    'name': 'storage.modifier.visibility.set.${2:/downcase}.php'
                  '3':
                    'name': 'storage.modifier.visibility.${3:/downcase}.php'
                  '4':
                    'name': 'storage.modifier.${4:/downcase}.php'
```

alexchexes replied:
same problem with word boundary after `set)`

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -126,8 +136,21 @@
+                      '3':
+                        'name': 'storage.modifier.visibility.${3:/downcase}.php'
+                      '4':
+                        'name': 'storage.modifier.static.php'
```
```suggestion
                    'match': '''(?xi)
                      \\b(?:
                        (final|abstract)
                        |(public|private|protected)
                        |(static)
                      )\\b
                    '''
                    'name': 'storage.modifier.php'
                    'captures':
                      '1':
                        'name': 'storage.modifier.inheritance.${1:/downcase}.php'
                      '2':
                        'name': 'storage.modifier.visibility.${2:/downcase}.php'
                      '3':
                        'name': 'storage.modifier.static.php'
```

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -860,8 +934,24 @@
+      '4':
+        'name': 'storage.modifier.visibility.${4:/downcase}.php'
+      '5':
+        'name': 'storage.modifier.static.php'
```
```suggestion
    'match': '''(?xi)
      \\b(?:
        (global)
        |(abstract|final)
        |(private|protected|public)
        |(static)
      )\\b
    '''
    'name': 'storage.modifier.php'
    'captures':
      '1':
        'name': 'storage.modifier.global.php'
      '2':
        'name': 'storage.modifier.inheritance.${2:/downcase}.php'
      '3':
        'name': 'storage.modifier.visibility.${3:/downcase}.php'
      '4':
        'name': 'storage.modifier.static.php'
```

alexchexes committed:
Apply suggestions from code review

Co-authored-by: KapitanOczywisty <44417092+KapitanOczywisty@users.noreply.github.com>

alexchexes committed:
reduce amount of capture groups

alexchexes commented:
@KapitanOczywisty Done. Some regexes could not be simplified in the proposed way, because we cannot add a word boundary after `)`. Otherwise, all changes have been applied

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -659,8 +692,19 @@
             'patterns': [
               {
-                'match': '(?:public|private|protected)(?:\\(set\\))?|readonly'
+                'match': '''(?x)
```
```suggestion
                'match': '''(?xi)
```

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -785,8 +840,19 @@
         'patterns': [
           {
-            'match': '(?:public|private|protected)(?:\\(set\\))?|static|readonly'
+            'match': '''(?x)
```
```suggestion
            'match': '''(?xi)
```

KapitanOczywisty commented on grammars/php.cson:
```diff
@@ -2753,8 +2835,10 @@
+          '5':
             'name': 'invalid.illegal.wrong-access-type.phpdoc.php'
-        'match': '^\\s*\\*\\s*(@access)\\s+((public|private|protected)|(.+))\\s*$'
+        'match': '^\\s*\\*\\s*(@access)\\s+(((public|private|protected))|(.+))\\s*$'
```
```suggestion
        'captures':
          '1':
            'name': 'keyword.other.phpdoc.php'
          '2':
            'name': 'storage.modifier.php'
          '3':
            'name': 'storage.modifier.visibility.${3:/downcase}.php'
          '4':
            'name': 'invalid.illegal.wrong-access-type.phpdoc.php'
        'match': '^\\s*\\*\\s*(@access)\\s+(?i:((public|private|protected))|(.+))\\s*$'
```

Just for consistency allow case-insensitive

alexchexes committed:
Apply suggestions from code review

to pull and verify locally

Co-authored-by: KapitanOczywisty <44417092+KapitanOczywisty@users.noreply.github.com>

alexchexes committed:
Add test coverage fox mixed-case modifiers. Remove duplicated capture block

alexchexes commented:
@KapitanOczywisty Done, also added test coverage
